src/cargo/sources/git/utils.rs

```diff
@@ -181,9 +181,13 @@ impl GitDatabase {
             .filter(|co| co.is_fresh())
         {
             Some(co) => co,
-            None => GitCheckout::clone_into(dest, self, rev, gctx)?,
+            None => {
+                let checkout = GitCheckout::clone_into(dest, self, rev, gctx)?;
+                checkout.update_submodules(gctx)?;
+                checkout
+            }
         };
-        checkout.update_submodules(gctx)?;
+
```

Comment:
`is_fresh` is determined by the presence of the `CHECKOUT_READY_LOCK` file.  That file is written in `reset` which is called at the end of `clone_into`.  As-written, if the submodule update fails or gets interrupted, we won't be able to detect that and recover on the next invocation.  We need to make sure that `CHECKOUT_READY_LOCK` is not written to until after the submodule update to make sure we can recover gracefully.